### ISSUE TITLE
Rendition constraint exceeds maximum MySQL index key length

### ISSUE THREAD

**Issue by chosak** [wagtail/wagtail#3256]
### Issue Summary

The change to image rendition model constraints [in the 1.8 release](http://docs.wagtail.io/en/v1.8/releases/1.8.html#unique-together-constraint-on-custom-image-rendition-models-needs-updating) causes problems with MySQL index key lengths. Specifically, adding this constraint:

```py
unique_together = (
    ('image', 'filter_spec', 'focal_point_key'),
)
```

creates this MySQL constraint command:

```sql
ALTER TABLE `renditionstable` ADD CONSTRAINT `indexname_uniq` UNIQUE (`image_id`, `filter_spec`, `focal_point_key`);
```

on a table like this:

```sql
mysql> DESCRIBE renditionstable;
+-----------------+--------------+------+-----+---------+----------------+
| Field           | Type         | Null | Key | Default | Extra          |
+-----------------+--------------+------+-----+---------+----------------+
| image_id        | int(11)      | NO   | MUL | NULL    |                |
| filter_spec     | varchar(255) | NO   | MUL | NULL    |                |
| focal_point_key | varchar(255) | NO   |     | NULL    |                |
| ...             |              |      |     |         |                |
+-----------------+--------------+------+-----+---------+----------------+
```

This fails on both a new database and an existing one with

```sql
ERROR 1071 (42000): Specified key was too long; max key length is 1000 bytes
```

This constraint is trying to create a `UNIQUE` index on three columns, two of which are type `varchar(255)`. When using the `utf8` encoding, these columns [use](http://dev.mysql.com/doc/refman/5.7/en/charset-unicode-utf8.html) 3 bytes per character (for `uf8mb4`, it's 4 bytes per).

When [creating an index](http://dev.mysql.com/doc/refman/5.7/en/create-index.html) in MySQL, "key values are formed by concatenating the values of the given columns." This means that the index width would have to be (11) + (255 * 3) + (255 *3), which exceeds the maximum key length of 1000 bytes (the limit for the MyISAM storage engine). In Wagtail 1.7 and earlier because the constraint was on `filter_id` instead, the index size would have been (11) + (11) + (255 *3), which fit okay under the limit.

Using the MySQL InnoDB storage engine (instead of MyISAM) this limit is 767 bytes by default although it can be increased to 3072 using [`--innodb-large-prefix`](http://dev.mysql.com/doc/refman/5.7/en/innodb-parameters.html#sysvar_innodb_large_prefix). This would likely satisfy this requirement, but it would be nice to be able to use MyISAM as well somehow.

One possible fix/workaround for this would have been to define `AbstractRendition.filter_spec` and `.focal_point_key` as something like `max_length=64` versus the current `255`. Assuming those columns could live with the narrower widths, this change would get this constraint under both the 767 and 1000 byte limits. On the other hand, I can see why it might be good to standardize/future-proof these fields, and it might be hard to retrofit this change given that 1.8 has already shipped.

Beyond that idea I'm not sure a good way to workaround this, short of moving to InnoDB (not preferred). Another idea might be to avoid use of `AbstractRendition` at all, and just duplicate needed functionality in a custom rendition class, with narrower columns as suggested above. Any thoughts?

(This migration was also mentioned by @bkfox in [#2953](https://github.com/wagtail/wagtail/issues/2953#issuecomment-269866098), but as that issue is more generally about index lengths, I thought it better to create a more specific one here. #2925 is also somewhat related.)

### Steps to Reproduce

1. Run a local MySQL instance using the MyISAM storage engine (`SET default_storage_engine=MYISAM`).
2. Create a Django project pointing to your MySQL, and then run migrations against Wagtail 1.8.
3. You should see this error when `wagtailimages.0016_deprecate_rendition_filter_relation` is run.

### Technical details

* Python version: 2.7.11
* Django version: 1.8.15
* Wagtail version: 1.8

**Comment by gasman** [comment]
Hi @chosak - thanks for the comprehensive report!

Please can you clarify something - if I follow your 'steps to reproduce' against a MySQL instance configured with MyISAM and utf8mb4, the migrations fail at wagtailimages.0001_initial with "Specified key was too long; max key length is 1000 bytes", which is what I would expect given #2925 and #2953. (If I use InnoDB, I get the same thing but "max key length is 767 bytes".) Is there some MySQL configuration or workaround you've applied that allows you to get past migration 0001, but subsequently fails at migration 0016? Or is this more of a "future" bug report that will surface once #2925 / #2953 are fixed?

**Comment by chosak** [comment]
@gasman thanks for following up about this.

I'm using a default character set of `utf8` for my DB (3 bytes per character), so I can get past the initial migration, just not this new change. Because `utf8mb4` uses 4 bytes per character, the `focal_point_key` part of that `0001_initial` constraint (max length 255) already exceeds the limit (255 * 4). Using `utf8`, you're still okay until migration `0016`.

I should have specified "with `utf8`" more prominently in my bug report -- sorry for any confusion.

**Comment by gasman** [comment]
Thanks! I can confirm I'm seeing that behaviour with MyISAM and `utf8` now. (I thought I'd tested against `utf8` before with no errors, but it seems I got the incantation wrong and it was still using `latin1`...)

I've made a tentative fix here https://github.com/gasman/wagtail/commit/96c18908a7de451c502874af35726d177a101bc1 - but since this is a breaking change that affects a previously-working setup, it warrants a 1.8.1 release, and so I'll need to reshuffle the migrations so that this comes between the current 0016 and 0017.

**Comment by chosak** [comment]
This seems like a good approach.

I tried your branch, and for an existing 1.7 deployment that has already had the old `wagtail_images.0004` migration, applying `wagtail_images.0016` still raises the above error for me. Could your new migration instead come before 0016, so that `focal_point_key` is guaranteed to be smaller before the unique_together constraint is [changed to include](https://github.com/wagtail/wagtail/blob/master/wagtail/wagtailimages/migrations/0016_deprecate_rendition_filter_relation.py#L20) `filter_spec`?

**Comment by gasman** [comment]
Yep, certainly - I realised the mistake just after posting the above :-)

Since migration 0016 has already shipped in Wagtail 1.8, we can't put anything before it (because that'll give people upgrading 1.8 -> 1.8.1 an inconsistent migration history error), so I think the appropriate workaround is to add the AlterField to the start of 0016 (so that a 1.7 -> 1.8.1 upgrade on MySQL succeeds), then do it _again_ as migration 0017 (to ensure that 1.8 -> 1.8.1 upgrades don't miss that step).

**Comment by chosak** [comment]
Thanks, that sounds great.

**Comment by gasman** [comment]
Hi @chosak - I've now pushed my proposed fix at #3293 (for 1.8.x) and #3294 (for 1.9.x / master) if you want to give it a try.

**Comment by chosak** [comment]
Thanks @gasman. I encountered a problem with #3293 and commented over there. Unfortunately I need to think more about a workaround for this - maybe adding _another_ migration that does a redundant field change and then change back, just to make sure it gets applied?

**Comment by gasman** [comment]
Argh, that's annoying - I didn't realise that exact duplicate AlterField operations would be optimised away. Perhaps we could fiddle the max_length on the earlier migrations so that it steps down through 18, 17, 16 - that way there's always a real database-level change to be made, regardless of where we pick up the migration sequence...?

**Comment by gasman** [comment]
Have updated the PRs to "step down" the max_length as proposed - seems to do the job.

Tested with the following configuration:

    DATABASES = {
        'default': {
            'ENGINE': 'django.db.backends.mysql',
            'NAME': 'utf8test',
            'USER': 'root',
            'PASSWORD': '...',
            'OPTIONS': {
                'charset': 'utf8',
                'init_command': "SET sql_mode='STRICT_TRANS_TABLES'; SET default_storage_engine=MYISAM; SET NAMES utf8;",
            }
        }
    }

and a database created with `mysql -uroot -p -e "CREATE DATABASE utf8test CHARACTER SET utf8 COLLATE utf8_general_ci"`. Performed the following tests:

1. Starting from a fresh install on 1.7, upgrade to the `fix/mysql-index-length-1.8` branch, run `./manage.py migrate`. Upgrade to the `fix/mysql-index-length-1.9` branch, run `./manage.py migrate`
2. Starting from a fresh install on `fix/mysql-index-length-1.8`, upgrade to the `fix/mysql-index-length-1.9` branch, run `./manage.py migrate`
3. Bring up a fresh install on `fix/mysql-index-length-1.9`
4. Starting from a fresh install on 1.7, upgrade to the `fix/mysql-index-length-1.9` branch, run `./manage.py migrate`

Also tested the upgrade from 1.8 -> `fix/mysql-index-length-1.8` -> `fix/mysql-index-length-1.9` and 1.8 -> `fix/mysql-index-length-1.9`, on a Postgres instance.

**Comment by gasman** [comment]
Have confirmed that the above also works with `SET default_storage_engine=INNODB;` in init_command.

**Comment by chosak** [comment]
:+1: I can verify that these migrations all pass for my `utf8` MyISAM setup now, including a custom image rendition model:

1.7 -> 1.8a0
1.7 -> 1.8a0 -> 1.9a0
1.7 -> 1.9a0

Thanks for the creative solution!

**Comment by gasman** [comment]
@chosak Thanks for the confirmation!

Now merged into 1.8.x and master.
